query-engine/core/src/query_graph_builder/write/upsert.rs

```diff
@@ -177,3 +191,73 @@ pub fn upsert_record(
 
     Ok(())
 }
+
+// This optimisation on our upserts allows us to use the `INSERT ... ON CONFLICT SET ..`
+// when the query matches the following conditions:
+// 1. The data connector supports it
+// 2. The create and update arguments do not have any nested queries
+// 3. There is only 1 unique field in the where clause
+// 4. The unique field defined in where clause has the same value as defined in the create arguments
+fn can_use_connector_native_upsert(
+    model: &ModelRef,
+    where_field: &ParsedInputMap,
+    create_argument: &ParsedInputMap,
+    update_argument: &ParsedInputMap,
+    connector_ctx: &ConnectorContext,
+) -> bool {
+    let has_nested_create = create_argument
+        .iter()
+        .any(|(field_name, _)| model.fields().find_from_relation_fields(&field_name).is_ok());
+
+    let has_nested_update = update_argument
+        .iter()
+        .any(|(field_name, _)| model.fields().find_from_relation_fields(&field_name).is_ok());
+
+    let has_one_unique = where_field
+        .iter()
+        .filter(|(field_name, input)| {
+            is_unique_field(field_name, model) && where_and_create_equal(field_name, input, &create_argument)
+        })
+        .count()
+        == 1;
+
+    connector_ctx.can_native_upsert() && has_one_unique && !has_nested_create && !has_nested_update
+}
+
+fn is_unique_field(field_name: &String, model: &ModelRef) -> bool {
+    match model.fields().find_from_scalar(&field_name) {
+        Ok(field) => field.unique(),
+        Err(_) => resolve_compound_field(field_name, model).is_some(),
+    }
+}
+
+/// Make sure the unique fields defined in the where clause have the same values
+/// as in the create of the upsert.
+fn where_and_create_equal(
+    field_name: &String,
+    where_input: &ParsedInputValue,
+    create_argument: &ParsedInputMap,
+) -> bool {
+    if let ParsedInputValue::Map(map) = where_input {
+        map.iter()
+            .all(|(field_name, map_input)| where_and_create_equal_internal(field_name, map_input, &create_argument))
+    } else {
+        where_and_create_equal_internal(field_name, where_input, &create_argument)
+    }
+}
+
+fn where_and_create_equal_internal(
+    field_name: &String,
+    where_input: &ParsedInputValue,
+    create_argument: &ParsedInputMap,
+) -> bool {
+    let arg = create_argument
+        .iter()
+        .find(|(create_name, _)| create_name == &field_name);
+
+    if let Some((_, create_input)) = arg {
+        create_input == where_input
+    } else {
+        false
+    }
+}
```

Comment:
I think both of those functions can be reduced to:

```rust
/// Make sure the unique fields defined in the where clause have the same values
/// as in the create of the upsert.
fn where_and_create_equal(where_map: &ParsedInputMap, create_map: &ParsedInputMap) -> bool {
    where_map.iter().all(|(k, v)| match v {
        ParsedInputValue::Map(inner_map) => where_and_create_equal(inner_map, create_map),
        _ => Some(v) == create_map.get(k),
    })
}
```

With the usage simplified to:

```rust
let where_values_same_as_create = where_and_create_equal(where_argument, create_argument);
```

Reply:
I see you've refactored to only do the check on unique fields. Then, things could look like so:

```rust
/// Make sure the unique fields defined in the where clause have the same values
/// as in the create of the upsert.
fn where_and_create_equal(field_name: &str, where_value: &ParsedInputValue, create_map: &ParsedInputMap) -> bool {
    match where_value {
        ParsedInputValue::Map(inner_map) => inner_map
            .iter()
            .all(|(inner_field, inner_value)| where_and_create_equal(inner_field, inner_value, create_map)),
        _ => Some(where_value) == create_map.get(field_name),
    }
}
```

To be clear, what's _most_ important here is to use the indexed properties of the create map instead of iterating over it.

Refactoring `where_and_create_internal` to:

```
fn where_and_create_equal_internal(
    field_name: &String,
    where_input: &ParsedInputValue,
    create_argument: &ParsedInputMap,
) -> bool {
  Some(where_input) == create_argument.get(field_name)
}
```

Would already be great. My approach just removes the need for a second function altogether.